<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>PCA II</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
</style>


<script src="pca_2_files/libs/clipboard/clipboard.min.js"></script>
<script src="pca_2_files/libs/quarto-html/quarto.js"></script>
<script src="pca_2_files/libs/quarto-html/popper.min.js"></script>
<script src="pca_2_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="pca_2_files/libs/quarto-html/anchor.min.js"></script>
<link href="pca_2_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="pca_2_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="pca_2_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="pca_2_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="pca_2_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">

  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

</head>

<body>

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">
<div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
  <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#data-variance-and-centering" id="toc-data-variance-and-centering" class="nav-link active" data-scroll-target="#data-variance-and-centering">Data Variance and Centering</a></li>
  <li><a href="#pca-review" id="toc-pca-review" class="nav-link" data-scroll-target="#pca-review">PCA Review</a>
  <ul>
  <li><a href="#using-principal-components" id="toc-using-principal-components" class="nav-link" data-scroll-target="#using-principal-components">Using Principal Components</a></li>
  </ul></li>
  <li><a href="#case-study-house-of-representatives-voting" id="toc-case-study-house-of-representatives-voting" class="nav-link" data-scroll-target="#case-study-house-of-representatives-voting">Case Study: House of Representatives Voting</a></li>
  <li><a href="#interpreting-pca" id="toc-interpreting-pca" class="nav-link" data-scroll-target="#interpreting-pca">Interpreting PCA</a>
  <ul>
  <li><a href="#pca-plot" id="toc-pca-plot" class="nav-link" data-scroll-target="#pca-plot">PCA Plot</a></li>
  <li><a href="#scree-plots" id="toc-scree-plots" class="nav-link" data-scroll-target="#scree-plots">Scree Plots</a></li>
  <li><a href="#pca-with-svd" id="toc-pca-with-svd" class="nav-link" data-scroll-target="#pca-with-svd">PCA with SVD</a></li>
  <li><a href="#columns-of-v-are-the-directions" id="toc-columns-of-v-are-the-directions" class="nav-link" data-scroll-target="#columns-of-v-are-the-directions">Columns of V are the Directions</a></li>
  <li><a href="#biplots" id="toc-biplots" class="nav-link" data-scroll-target="#biplots">Biplots</a></li>
  </ul></li>
  <li><a href="#applications-of-pca" id="toc-applications-of-pca" class="nav-link" data-scroll-target="#applications-of-pca">Applications of PCA</a>
  <ul>
  <li><a href="#pca-in-biology" id="toc-pca-in-biology" class="nav-link" data-scroll-target="#pca-in-biology">PCA in Biology</a></li>
  <li><a href="#image-classification" id="toc-image-classification" class="nav-link" data-scroll-target="#image-classification">Image Classification</a></li>
  <li><a href="#why-pca-then-model" id="toc-why-pca-then-model" class="nav-link" data-scroll-target="#why-pca-then-model">Why PCA, then Model?</a></li>
  </ul></li>
  </ul>
</nav>
</div>
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">PCA II</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<section id="data-variance-and-centering" class="level2">
<h2 class="anchored" data-anchor-id="data-variance-and-centering">Data Variance and Centering</h2>
<p>We define the total variance of a data matrix as the sum of variances of attributes.</p>
<p>Formally, the <span class="math inline">\(i\)</span>-th singular value tells us the <strong>component score</strong>, i.e., how much of the data variance is captured by the ith principal component. Supposing the number of data points is <span class="math inline">\(n\)</span>:</p>
<p><span class="math display">\[\text{i-th component score} = \frac{(\text{i-th singular value}^2)}{n}\]</span></p>
<p>Summing up the component scores is equivalent to computing total variance.</p>
<p><strong>Data Centering</strong>: PCA has a data centering step that precedes any singular value decomposition, where if implemented defines the component score as above.</p>
<div class="callout callout-style-default callout-note no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-1-contents" aria-controls="callout-1" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Learning Outcomes
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-1" class="callout-1-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<ul>
<li>Develop a deeper understanding of how to interpret Principal Components Analysis (PCA).</li>
<li>See applications of PCA to some real-world contexts.</li>
</ul>
</div>
</div>
</div>
</section>
<section id="pca-review" class="level2">
<h2 class="anchored" data-anchor-id="pca-review">PCA Review</h2>
<section id="using-principal-components" class="level3">
<h3 class="anchored" data-anchor-id="using-principal-components">Using Principal Components</h3>
<p>Steps to obtain Principal Components via SVD:</p>
<ol type="1">
<li><p>Center the data matrix by subtracting the mean of each attribute column.</p></li>
<li><p>To find the <span class="math inline">\(k\)</span> <strong>principal components</strong>:</p></li>
</ol>
<ul>
<li>Compute the SVD of the data matrix (<span class="math inline">\(X = U{\Sigma}V^{T}\)</span>)</li>
<li>The first <span class="math inline">\(k\)</span> columns of <span class="math inline">\(U{\Sigma}\)</span> (or equivalently, <span class="math inline">\(XV\)</span>) contain the <span class="math inline">\(k\)</span> principal components of <span class="math inline">\(X\)</span>.</li>
</ul>
<p>The principal components are a low-dimension representation that capture as much of the original data’s total variance as possible.</p>
<p>The <strong>component scores</strong> sum to total variance if we center our data. <span class="math display">\[\text{component score}_i = \frac{\sigma_i^{2}}{N}\]</span></p>
<p>We can also use the SVD to get a rank-<span class="math inline">\(k\)</span> approximation of <span class="math inline">\(X\)</span>, <span class="math inline">\(X_k\)</span>.</p>
<p><span class="math display">\[X_k = \sum_{j = 1}^{k} \sigma_ju_jv_j^{T} \]</span></p>
<p>where <span class="math inline">\(\sigma_j\)</span> is the <span class="math inline">\(j\)</span>-th singular value of <span class="math inline">\(X\)</span>, <span class="math inline">\(u_j\)</span> is the <span class="math inline">\(j\)</span>-th column of <span class="math inline">\(U\)</span>, and <span class="math inline">\(v_j\)</span> is the <span class="math inline">\(j\)</span>-th column of <span class="math inline">\(V\)</span>.</p>
<p>If you want to dive deeper into PCA, <a href="https://www.youtube.com/playlist?list=PLMrJAkhIeNNSVjnsviglFoY2nXildDCcv">Steve Brunton’s SVD Video Series</a> is a great resource.</p>
</section>
</section>
<section id="case-study-house-of-representatives-voting" class="level2">
<h2 class="anchored" data-anchor-id="case-study-house-of-representatives-voting">Case Study: House of Representatives Voting</h2>
<p>Let’s examine how the House of Representatives (of the 116th Congress, 1st session) voted in the month of September 2019.</p>
<p>Specifically, we’ll look at the records of Roll call votes. From the U.S. Senate (<a href="https://www.senate.gov/reference/Index/Votes.htm">link</a>):</p>
<ul>
<li>Roll call votes occur when a representative or senator votes “yea” or “nay,” so that the names of members voting on each side are recorded. A voice vote is a vote in which those in favor or against a measure say “yea” or “nay,” respectively, without the names or tallies of members voting on each side being recorded.</li>
</ul>
<p><strong>Do legislators’ roll call votes show a relationship with their political party?</strong></p>
<p>Please visit this <a href="https://data100.datahub.berkeley.edu/hub/user-redirect/git-pull?repo=https%3A%2F%2Fgithub.com%2FDS-100%2Ffa23-student&amp;urlpath=lab%2Ftree%2Ffa23-student%2Flecture%2Flec26%2Flec26-votes.ipynb&amp;branch=main">link</a> to see the full Jupyter notebook demo.</p>
<p>As shown in the demo, the primary goal of PCA is to transform observations from high-dimensional data down to low dimensions through linear transformations.</p>
<p>A related goal of PCA relates to the idea that a low-dimension representation of the data should capture the variability of the original data. For example, if the first two singular values are large and the others are relatively small, then two dimensions are probably enough to describe most of waht distinguishes one observation from another. However, if this is not the case, then a PCA scatter plot is probably omitting lots of information.</p>
<p>We can use the the following formulas to quantify the amount each prinicial component contributes to the total variance:</p>
<p><span class="math display">\[ \text{component score} = \frac{\sigma_i^{2}}{N}\]</span></p>
<p><span class="math display">\[ \text{total variance} = \text{sum of all the component scores} = \sum_{i=1}^k \frac{\sigma_i^{2}}{N} \]</span></p>
<p><span class="math display">\[ \text{variance ratio of principal component i} = \frac{\text{component score i}}{\text{total variance}} = \frac{\sigma_i^{2} / N}{\sum_{i=1}^k \sigma_i^{2} / N}\]</span></p>
<p>In Python, assuming you had a 1D <code>NumPy</code> array of singular values <code>s</code> returned by <code>np.linalg.svd</code>, you could compute the list of variances ratios with <code>s**2 / sum(s**2)</code>.</p>
</section>
<section id="interpreting-pca" class="level2">
<h2 class="anchored" data-anchor-id="interpreting-pca">Interpreting PCA</h2>
<section id="pca-plot" class="level3">
<h3 class="anchored" data-anchor-id="pca-plot">PCA Plot</h3>
<p>We often plot the first two principal components using a scatter plot, with PC1 on the <span class="math inline">\(x\)</span>-axis, PC2 on the <span class="math inline">\(y\)</span>-axis. This is often called a PCA plot.</p>
<p>If the first two singular values are large and all others are small, then two dimensions are enough to describe most of what distinguishes one observation from another. If not, then a PCA plot is omitting lots of information.</p>
<p>PCA plots help us assess similarities between our data points and if there are any clusters in our dataset. In the case study before, for example, we could create the following PCA plot:</p>
<p><img src="images/pca_plot.png" alt="pca_plot" width="500"></p>
</section>
<section id="scree-plots" class="level3">
<h3 class="anchored" data-anchor-id="scree-plots">Scree Plots</h3>
<p>A scree plot shows the <strong>variance ratio</strong> captured by each principal component, with the largest variance ratio first.</p>
<p>Scree plots help us visually determine the number of dimensions needed to describe the data reasonably. The singular values that fall in region of the plot after a large drop off correspond to principal components that are <strong>not</strong> needed to describe the data, since they explain a relatively low proportion of the total variance of the data. For example, in the below plot, we could use the “elbow method” just described to figure out that the first 2 PCs capture the bulk of the information.</p>
<p><img src="images/scree_plot.png" alt="scree_plot" width="500"></p>
</section>
<section id="pca-with-svd" class="level3">
<h3 class="anchored" data-anchor-id="pca-with-svd">PCA with SVD</h3>
<p>After finding the SVD of <span class="math inline">\(X\)</span>:</p>
<p><img src="images/slide15.png" alt="slide15" width="500"></p>
<p>We can derive the principal components of the data. Specifically, the first <span class="math inline">\(n\)</span> rows of <span class="math inline">\(V^{T}\)</span> are directions for the <span class="math inline">\(n\)</span> principal components.</p>
</section>
<section id="columns-of-v-are-the-directions" class="level3">
<h3 class="anchored" data-anchor-id="columns-of-v-are-the-directions">Columns of V are the Directions</h3>
<p><img src="images/slide16.png" alt="slide16" width="500"></p>
<p>The elements of each column of <span class="math inline">\(V\)</span> (row of <span class="math inline">\(V^{T}\)</span>) rotate the original feature vectors into a principal component.</p>
<p>The first column of V indicates how each feature contributes (e.g.&nbsp;positive, negative, etc.) to principal component 1.</p>
<p><img src="images/rotate_center_plot.png" alt="slide17" width="750"></p>
<p>Coupled together, this interpretation also allows us to understand that:</p>
<ol type="1">
<li>The Principal components are all <strong>orthogonal</strong> to each other because the columns of U are orthonormal.</li>
<li>Principal Components are <strong>axis-aligned</strong>. That is, if you plot two PCs on a 2D plane, one will lie on the x-axis, the other on the y-axis.</li>
<li>Principal Components are <strong>linear combinations</strong> of columns in our data X</li>
</ol>
</section>
<section id="biplots" class="level3">
<h3 class="anchored" data-anchor-id="biplots">Biplots</h3>
<p>Biplots superimpose the directions onto the plot of PC2 vs PC1.</p>
<p>Vector <span class="math inline">\(j\)</span> corresponds to the direction for feature <span class="math inline">\(j\)</span> (e.g.&nbsp;<span class="math inline">\(v_{1j}, v_{2j}\)</span>). - There are several ways to scale biplots vectors; in this course we plot the direction itself. - For other scalings, which can lead to more interpretable directions/loadings, see <a href="https://blogs.sas.com/content/iml/2019/11/06/what-are-biplots.html">SAS biplots</a></p>
<p>Through biplots, we can interpret how features correlate with the principal components shown: positively, negatively, or not much at all.</p>
<p><img src="images/slide17_2.png" alt="slide17_2" width="500"></p>
<p>The directions of the arrow are (<span class="math inline">\(v_1\)</span>, <span class="math inline">\(v_2\)</span>) where <span class="math inline">\(v_1\)</span> and <span class="math inline">\(v_2\)</span> are how that specific feature column contributes to PC1 and PC2, respectively. <span class="math inline">\(v_1\)</span> and <span class="math inline">\(v_2\)</span> are elements of the first and second columns of <span class="math inline">\(V\)</span>, respectively (i.e., the first two rows of <span class="math inline">\(V^T\)</span>).</p>
<p>Say we were considering feature 3, and say that was the green arrow here (pointing bottom right).</p>
<ul>
<li><p><span class="math inline">\(v_1\)</span> and <span class="math inline">\(v_2\)</span> are the third elements of the respective columns in <span class="math inline">\(V\)</span>. They are what scale feature 3’s column vector in the linear transformation to PC1 and PC2, respectively.</p></li>
<li><p>Here we would infer that <span class="math inline">\(v_1\)</span> (in the <span class="math inline">\(x\)</span>/PC1-direction) is positive, meaning that a linear increase in feature 3 would correspond to a linear increase of PC1, meaning feature 3 and PC1 are positively correlated.</p></li>
<li><p><span class="math inline">\(v_2\)</span> (in the <span class="math inline">\(y\)</span>/pc2-direction) is negative, meaning a linear increase in feature 3 would result correspond to a linear decrease in PC2, meaning feature 3 and PC2 are negatively correlated.</p></li>
</ul>
</section>
</section>
<section id="applications-of-pca" class="level2">
<h2 class="anchored" data-anchor-id="applications-of-pca">Applications of PCA</h2>
<section id="pca-in-biology" class="level3">
<h3 class="anchored" data-anchor-id="pca-in-biology">PCA in Biology</h3>
<p>PCA is commonly used in biomedical contexts, which have many named variables!</p>
<ol type="1">
<li><p>To cluster data (<a href="https://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-019-2680-1">Paper 1</a>, <a href="https://www.science.org/doi/10.1126/scirobotics.abk2378">Paper 2</a>)</p></li>
<li><p>To identify correlated variables (<a href="https://docs.google.com/presentation/d/1-aDu0ILCkPx3iCcJGB3YXci-L4g90Q6AarXU6wffLB8/edit#slide=id.g62cb86badb_0_1128">interpret</a> rows of <span class="math inline">\(V^{T}\)</span> as linear coefficients) (<a href="https://www.nature.com/articles/s41598-017-05714-1">Paper 3</a>). Uses <a href="https://www.google.com/url?q=https://www.geo.fu-berlin.de/en/v/soga/Geodata-analysis/Principal-Component-Analysis/principal-components-basics/Interpretation-and-visualization/index.html%23:~:text%3DThe%2520biplot%2520is%2520a%2520very,in%2520a%2520single%2520biplot%2520display.%26text%3DThe%2520plot%2520shows%2520the%2520observations,principal%2520components%2520(synthetic%2520variables).&amp;sa=D&amp;source=editors&amp;ust=1682131633152964&amp;usg=AOvVaw2H9SOeMP5kUS890Fkhfthx">biplots</a>.</p></li>
</ol>
</section>
<section id="image-classification" class="level3">
<h3 class="anchored" data-anchor-id="image-classification">Image Classification</h3>
<p>In machine learning, PCA is often used as a preprocessing step prior to training a supervised model.</p>
<p>See the following <a href="https://data100.datahub.berkeley.edu/hub/user-redirect/git-pull?repo=https%3A%2F%2Fgithub.com%2FDS-100%2Ffa23-student&amp;urlpath=lab%2Ftree%2Ffa23-student%2Flecture%2Flec26%2Flec26-fashion-mnist.ipynb&amp;branch=main">demo</a> to see how PCA is useful for building an image classification model based on the MNIST-Fashion dataset.</p>
<p>The demo shows how we can use PCA during the Exploratory Data Analysis stage of our data science lifecycle to: - visually identify clusters of similar observations in high dimensions. - find a small basis for representing variations in complex things. - reduce the number of dimensions to make some computation cheaper.</p>
</section>
<section id="why-pca-then-model" class="level3">
<h3 class="anchored" data-anchor-id="why-pca-then-model">Why PCA, then Model?</h3>
<ol type="1">
<li>Reduces dimensionality, allowing us to speed up training and reduce the number of features, etc.</li>
<li>Avoids multicollinearity in the new features created (i.e.&nbsp;the principal components)</li>
</ol>
<p><img src="images/slide21.png" alt="slide21" width="500"></p>
</section>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>